stablehlo/dialect/TypeInference.cpp

```diff
@@ -1344,46 +1361,37 @@ LogicalResult inferAllToAllOp(
   return success();
 }
 
-LogicalResult inferBatchNormGradOp(
-    Optional<Location> location, Value operand, Value scale,
-    int64_t featureIndex,
-    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
-  if (failed(verifyBatchNorm(location, operand, scale, featureIndex)))
+LogicalResult inferBatchNormOp(
+    Optional<Location> location, ValueRange multiDimOperands,
+    ValueRange singleDimOperands, int64_t featureIndex,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes,
+    bool is_inference) {
+  if (failed(verifyBatchNorm(location, multiDimOperands, singleDimOperands,
+                             featureIndex)))
     return failure();
-  auto operandType = operand.getType().cast<RankedTensorType>();
-  inferredReturnShapes.emplace_back(operandType.cast<ShapedType>());
 
-  const int64_t featureCount = operandType.getDimSize(featureIndex);
-  SmallVector<int64_t> featureShape{featureCount};
-  inferredReturnShapes.emplace_back(featureShape, operandType.getElementType());
-  inferredReturnShapes.emplace_back(featureShape, operandType.getElementType());
-  return success();
-}
+  // Batch norm ops require operands to be ranked.
+  auto operandType = multiDimOperands[0].getType().cast<RankedTensorType>();
+  inferredReturnShapes.emplace_back(
+      operandType.getShape(), operandType.getElementType(),
+      operandType.cast<RankedTensorType>().getEncoding());
 
-LogicalResult inferBatchNormInferenceOp(
-    Optional<Location> location, Value operand, Value scale,
-    int64_t featureIndex,
-    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
-  if (failed(verifyBatchNorm(location, operand, scale, featureIndex)))
-    return failure();
-  auto operandType = operand.getType().cast<RankedTensorType>();
-  inferredReturnShapes.emplace_back(operandType.cast<ShapedType>());
-  return success();
-}
+  if (is_inference) return success();
 
-LogicalResult inferBatchNormTrainingOp(
-    Optional<Location> location, Value operand, Value scale,
-    int64_t featureIndex,
-    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
-  if (failed(verifyBatchNorm(location, operand, scale, featureIndex)))
-    return failure();
-  auto operandType = operand.getType().cast<RankedTensorType>();
-  inferredReturnShapes.emplace_back(operandType.cast<ShapedType>());
+  SmallVector<int64_t> featureShape{operandType.getDimSize(featureIndex)};
```

Comment:
`singleDimShape`, to be consistent with the multi-dim and single-dim nomenclature? (Same for `operandBounds` => `multiDimBounds`, `featureBounds` => `singleDimBounds`, `singleDimensionResultShape` => `singleDimShape`).